scripts/deploy-tarball.js

```diff
@@ -0,0 +1,153 @@
+    JSON.stringify({
+      name: 'files',
+      dependencies: {},
+      scripts: {
+        build: 'node inject-deploy-url.js',
+      },
+    })
+  )
+  await fs.copyFile(
+    path.join(cwd, 'scripts/inject-deploy-url.js'),
+    path.join(deployDir, 'inject-deploy-url.js')
+  )
+
+  let nativePackagesDir = path.join(cwd, 'packages/next-swc/crates/napi/npm')
+  let platforms = (await fs.readdir(nativePackagesDir)).filter(
+    (name) => !name.startsWith('.')
+  )
+
+  const optionalDeps = {}
+  const { version } = JSON.parse(
+    await fs.readFile(path.join(cwd, 'lerna.json'))
+  )
+
+  await Promise.all(
+    platforms.map(async (platform) => {
+      let binaryName = `next-swc.${platform}.node`
+      await fs.cp(
+        path.join(cwd, 'packages/next-swc/native', binaryName),
+        path.join(nativePackagesDir, platform, binaryName)
+      )
+      let pkg = JSON.parse(
+        await fs.readFile(
+          path.join(nativePackagesDir, platform, 'package.json')
+        )
+      )
+      pkg.version = version
+      await fs.writeFile(
+        path.join(nativePackagesDir, platform, 'package.json'),
+        JSON.stringify(pkg, null, 2)
+      )
+      const { stdout } = await execa(`npm`, [
+        `pack`,
+        `${path.join(nativePackagesDir, platform)}`,
+      ])
+      process.stdout.write(stdout)
+      const tarballName = stdout.split('\n').pop().trim()
+      await fs.rename(
+        path.join(cwd, tarballName),
+        path.join(publicDir, tarballName)
+      )
+      optionalDeps[pkg.name] = `https://DEPLOY_URL/${tarballName}`
+    })
+  )
+
+  const nextPkgJsonPath = path.join(cwd, 'packages/next/package.json')
+  const nextPkg = JSON.parse(await fs.readFile(nextPkgJsonPath, 'utf8'))
+
+  nextPkg.optionalDependencies = optionalDeps
+
+  await fs.writeFile(nextPkgJsonPath, JSON.stringify(nextPkg, null, 2))
+
+  const { stdout: nextPackStdout } = await execa(`npm`, [
+    `pack`,
+    `${path.join(cwd, 'packages/next')}`,
+  ])
+  process.stdout.write(nextPackStdout)
+  const nextTarballName = nextPackStdout.split('\n').pop().trim()
+  await fs.rename(
+    path.join(cwd, nextTarballName),
+    path.join(publicDir, nextTarballName)
+  )
+
+  await fs.writeFile(
+    path.join(deployDir, 'vercel.json'),
+    JSON.stringify(
+      {
+        version: 2,
+        rewrites: [
+          {
+            source: '/next.tgz',
+            destination: `/${nextTarballName}`,
+          },
+        ],
+      },
+      null,
+      2
+    )
+  )
+  const vercelConfigDir = path.join(cwd, '.vercel')
+
+  if (process.env.VERCEL_TEST_TOKEN) {
+    await fs.mkdir(vercelConfigDir)
+    await fs.writeFile(
+      path.join(vercelConfigDir, 'auth.json'),
+      JSON.stringify({
+        token: process.env.VERCEL_TEST_TOKEN,
+      })
+    )
+    await fs.writeFile(
+      path.join(vercelConfigDir, 'config.json'),
+      JSON.stringify({})
+    )
+    console.log('wrote config to', vercelConfigDir)
+  }
```

Comment:
You shouldn't need to write to a directory here. You can use `--token` flag.

Reply:
Take a look at this other script we use to deploy examples:

https://github.com/vercel/next.js/blob/d4069559f2134d6c0976d902beebd70e35bde0cb/scripts/deploy-examples.sh#L22-L23

Reply:
Token flag is risky as it exposes it in the logs and we have to hope GitHub logs hides it

Reply:
I thought github hides their own secrets so it should be fine. Otherwise that deploy-examples.sh script is going to have to be changed (and probably every other vercel cli script out there)

Maybe we can ask the CLI team if there is an env var we can use instead, since this is very common pattern and we don't want to depend on writing to a directory that may change in the future.

Reply:
The directory flag is explicitly specified so it would be a breaking change for that part to change across versions